data/src/main/java/org/apache/iceberg/data/orc/GenericOrcWriters.java

```diff
@@ -231,7 +231,24 @@ public void nonNullWrite(int rowId, String data, ColumnVector output) {
 
     @Override
     public void nonNullWrite(int rowId, ByteBuffer data, ColumnVector output) {
-      ((BytesColumnVector) output).setRef(rowId, data.array(), 0, data.array().length);
+      // We technically can't be sure if the ByteBuffer coming in is on or off
+      // heap so we cannot safely call `.array()` on it without first checking
+      // via the method ByteBuffer.hasArray().
+      // See: https://errorprone.info/bugpattern/ByteBufferBackingArray
+      //
+      // When there is a backing heap based byte array, we avoided the overhead of
+      // copying, which is especially important for small byte buffers.
+      //
+      // TODO - This copy slows it down, perhap unnecessarily. Is there any other way to tell, or no?
```

Comment:
I believe the write can only come from org.apache.orc.storage.ql.exec.vector.BytesColumnVector, which would be heap array backed. Unless there is another implementation I don't know about

Reply:
Isn't this writing _into_ a `BytesColumnVector`? If so, I don't think that the values are coming from one.

I think it is correct to updating the handling here.

It would be great to get an opinion from @shardulm94 and @edgarRd as well.

Reply:
Ah that is true, I was thinking about it the wrong direction

Reply:
I went looking into spark's implementation of off-heap byte arrays (for the spark-native version) and it somewhat lead me to believe that it could be reading in values as off-heap byte buffers when using `spark.memory.offHeap.enabled=true` and some value for  --conf `spark.memory.offHeap.size`.

I'm working on a demo to see if that's true, but of note when I started up a `spark-shell` with those parameters specified (`4G` for the off heap size), I got the following error which leads me to believe that ByteBuffers were being created in an off-heap way.

```
WARNING: Illegal reflective access by org.apache.spark.unsafe.Platform (file:/usr/local/Cellar/apache-spark/3.0.0/libexec/jars/spark-unsafe_2.12-3.0.0.jar) to constructor java.nio.DirectByteBuffer(long,int)
```
I am trying to work on a demo to test writing specifically off heap byte arrays to `format("orc")` and then try reading it in and see what happens, but I'm not 100% sure when I'll have the time.

But if the ByteBuffers are off heap and we pass them to this function without "sanitizing" them, we need the check. The check is pretty inexpensive as well fwiw, it just checks if a primitive array element if equal to null or not.

Reply:
Right, I agree that when writing this `ByteBuffer` to a `BytesColumnVector` there was an assumption that it was backed by an array, but to support any `ByteBuffer` we need this handling here.